runtime/bundles/org.eclipse.core.jobs/src/org/eclipse/core/runtime/jobs/Job.java

```diff
@@ -654,6 +732,72 @@ public final void schedule() {
 		super.schedule(0L);
 	}
 
+	/**
+	 * Schedule the job and returns a {@link CompletableFuture} that can be used to
+	 * observe the job state:
+	 * <ul>
+	 * <li>If the job is canceled, the future will be canceled and vice versa</li>
+	 * <li>When the job completes with an CANCEL status the future will be
+	 * canceled</li>
+	 * <li>When the job completes with an ERROR status the future will complete
+	 * exceptionally with a {@link CoreException}</li>
+	 * <li>In all other cases the future is completed with the job execution
+	 * status.</li>
+	 * </ul>
+	 *
+	 * @see #schedule()
+	 * @since 3.14
+	 */
+	public final CompletableFuture<IStatus> future() {
```

Comment:
I see that the JavaDoc says that `future` will also `schedule` but I don't feel like `future` as a method name conveys that very well. 

Something else:
Assuming I've an existing Job that was already running for some reason:
- future() will reschedule successfully
- the scheduled jobChangeEvent will be received
- the already running job will finish eventually
- the listener will be notified and the future will be completed

Yet this was not the run that was scheduled and cancelling the future would have cancelled to job that was already running and not the execution that was scheduled. In other words: `Job.future().cancel()` does not imply that the Job was cancelled.
Should we guard against this and only allow `future()` to be called if the job wasn't scheduled yet?

Reply:
we can name is `scheduleFuture()` if that's better, but I often prefer shorter names as it makes things easier to read.

About the other notes: Yes, because of how the Job API works there is no way to give hard guarantees, but that's already with the "traditional" Job usage, so the "owner" (the one how creates the job) is somewhat responsible how the job is used (e.g. do not share the instance with other threads that do perform concurrent schedules).

I previously have had tried to add a guard that one can only call `future()` if job is somehow "idle" state but found that even then there is no safe way to do so as the job might be scheduled right after this check... or just complete a microsecond afterwards and so on... The crucial part is really the adding of the listener and the scheduling of the job, so what one actually want is some kind of

`Job.trySchedule(0, myListener)`

and that listener is then always called only for that scheduled execution ... but this seems not be possible right now without major efforts.